op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+package db
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"slices"
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	entrySize                 = 24
+	searchCheckpointFrequency = 256
+
+	eventFlagIncrementLogIdx = byte(1)
+	//eventFlagHasExecutingMessage = byte(1) << 1
+)
+
+const (
+	typeSearchCheckpoint byte = iota
+	typeCanonicalHash
+	typeInitiatingEvent
+	typeExecutingLink
+	typeExecutingCheck
+)
+
+var (
+	ErrLogOutOfOrder  = errors.New("log out of order")
+	ErrDataCorruption = errors.New("data corruption")
+)
+
+type TruncatedHash []byte
```

Comment:
I think this, as a fixed-size 20 byte array, is more efficient (a Go slice header alone is 24 bytes) and easier for comparison (array ==).